### PR TITLE
added login to retrieve cookies instead of auth.py; 

### PR DESCRIPTION
**Pull request by ericdaltman**
Fully automated the script so no editing is needed by the user. It will allow you to enter in your username and password, then will grab the cookies and use them for the rest of the script functions in the same fashion as before, when they were hardcoded.

Compile tested and works great with pyinstaller and nuitka, coming in around 12MB on Linux and just under 10MB on Window.

### CODE REVIEW

**Comment by arty-name** [comment]
Wow, the power of open source! I am grateful to you, Eric, for this significant contribution.

Since everyone has different preferences, would you like a review from me and make the changes by your own, or would you rather have me add updates directly?

**Comment by ericdaltman** [comment]
Honestly, while I manage a dev team for work for years now, I have never contributed to a public git until now and really don't manage the actual codebase for my team... so you're kind of a first for me! The fun part was mapping out the communication needed using curl. I then subshelled the commands to test and finally converted those to requests. Then cleaned up the removal of auth.py trying to do minimal "damage" to the code already existing.

This is what happens when you remember you have a LiveJournal, and LiveJournal is owned pretty much by Putin, after work on a Friday night.

All that said? It's up to your preference, but I'm happy to do a review and change. Help me work out my python muscles a bit.

**Comment by arty-name** [comment]
The GitHub was showing me this is your first PR, but I thought it a mistake. Welcome!

File: download_comments.py [hunk 1]
```diff
@@ -4,10 +4,9 @@
 import json
 import requests
 import xml.etree.ElementTree as ET
-from auth import cookies, headers
 
-
-def fetch_xml(params):
+# Adding cookies and headers to definition for scope
```
Comment (arty-name) [inline download_comments.py]: These comments are helpful in the PR itself to indicate the reason for this change. When looking at the code later on, they would not make much sense. I guess let’s do without them.

File: download_comments.py [hunk 2]
```diff
@@ -72,12 +72,14 @@ def get_more_comments(start_id, users):
 
     return local_max_id, comments
 
-
-def download_comments():
+# Adding cookies and headers to definition since we won't get from auth.py
+def download_comments(Cookies, Headers):
+    cookies = Cookies
+    headers = Headers
```
Comment (arty-name) [inline download_comments.py]:
```suggestion
def download_comments(cookies, headers):
```
Would this also work the same way as in other functions?

File: download_posts.py [hunk 3]
```diff
@@ -1,16 +1,23 @@
...
+from sys import exit as sysexit
 import xml.etree.ElementTree as ET
-from auth import cookies, headers
 
+# At original import time, first request range of years. We'll add 1 to the end year automatically.
+try:
+    startYear = int(input("Enter first year you want to export: "))
+    endYear = int(input("Enter last year you want to export: "))
+except:
+    print("\nI'm guessing you didn't enter a whole number there, did ya? Wanna try again? Exiting...")
```
Comment (arty-name) [inline download_posts.py]: Since both inputs are wrapped in try/except it would not be clear to the user which of them caused the error. They will also miss potentially critical information on what exactly the error is, since the actual exception is not logged. Do you think it would make sense to have a small function which asks for a single integer and validates it? Also, English is not my first language, but would it be "the first"/"the last"?

File: download_posts.py [hunk 4]
```diff
@@ -52,15 +58,18 @@ def f(field):
         'current_mood': f('current_mood')
     }
 
-
-def download_posts():
+# Added cookies and headers to the definition as they'll now be passed by the main python script
+def download_posts(cookies, headers):
+    cookies = cookies
+    headers = headers
```
Comment (arty-name) [inline download_posts.py]: Do these two lines look pointless to you?

File: download_posts.py [hunk 5]
```diff
@@ -52,15 +58,18 @@ def f(field):
...
     os.makedirs('posts-xml', exist_ok=True)
     os.makedirs('posts-json', exist_ok=True)
 
     xml_posts = []
     for year in YEARS:
         for month in range(1, 13):
-            xml = fetch_month_posts(year, month)
+            xml = fetch_month_posts(year, month, cookies, headers)
+            #print("XML content:", xml)
```
Comment (arty-name) [inline download_posts.py]:
```suggestion
            
```
Do we need to keep this code?

File: export.py [hunk 6]
```diff
@@ -1,16 +1,74 @@
...
+    "Upgrade-Insecure-Requests": "1",
+    "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/127.0.0.0 Safari/537.36 OPR/113.0.0.0",
+    "sec-ch-ua": '"Chromium";v="127"',
+    "sec-ch-ua-platform": '"Windows"',
+}
+
+try:
+    # Get a "luid" cookie so it'll accept our form login.
+    cookies = {
+        'luid': requests.get("https://www.livejournal.com/", headers=headers).headers.get('Set-Cookie').split('luid=')[1].split(';')[0]
```
Comment (arty-name) [inline export.py]: This is one of three places where we get the value of a cookie from a string, does this deserve a dedicated function?

File: export.py [hunk 7]
```diff
@@ -1,16 +1,74 @@
...
+    }
+
+except:
+    # If attempt to reach LiveJournal fails, error out.
+    print("Couldn't reach www.livejournal.com, please check internet connection. Exiting.")
```
Comment (arty-name) [inline export.py]: Here we lose the information about the exact problem, would it be helpful to keep it?

File: export.py [hunk 8]
```diff
@@ -1,16 +1,74 @@
...
+
+# Login with user credentials and retrieve the two cookies required for the main script functions
+response = requests.post("https://www.livejournal.com/login.bml", data=credentials, cookies=cookies)
+
+# If not successful, whine about it.
+if response.status_code != 200:
+    print("Error - Return code:", response.status_code)
+
+# If successful, then get the 'Set-Cookie' key from the headers dict and parse it for the two cookies, placing them in a cookies dict
+curlOUTPUT = response.headers.get('Set-Cookie')
```
Comment (arty-name) [inline export.py]: Since this is not related to curl anymore, should the name be different?

File: export.py [hunk 9]
```diff
@@ -1,16 +1,74 @@
...
+}
+
+# Credit to the Author!
+headers = {
+    'User-Agent': 'https://github.com/arty-name/livejournal-export; me@arty.name'
+}
+
+# If a cookie could not be successfully acquired, error out. Otherwise, get to work.
+if any(value is None for value in cookies.values()):
+    print("Could not extract cookie data despite successful login.")
```
Comment (arty-name) [inline export.py]: Similar to years handling, two similar operations could fail and we won’t know which. Having a function to get and validate value could help here, what do you think?

File: export.py [hunk 10]
```diff
@@ -95,19 +142,14 @@ def json_to_markdown(json):
 
 def group_comments_by_post(comments):
     posts = {}
-
```
Comment (arty-name) [inline export.py:98]: I value my newlines, they highlight the structure of the code. Removing them turns the structure into a wall of text, please undo this.

File: export.py [hunk 11]
```diff
@@ -189,11 +226,11 @@ def combine(all_posts, all_comments):
         save_as_html(id, subfolder, json_post, post_comments_html)
         save_as_markdown(id, subfolder, json_post, post_comments_html)
 
-
 if __name__ == '__main__':
     if True:
-        all_posts = download_posts()
-        all_comments = download_comments()
+        # modified these calls to include cookies and headers since they can't be imported from auth.py anymore
```
Comment (arty-name) [inline export.py]:
```suggestion
        # modified these calls to include cookies and headers since they can't be imported from auth.py anymore
```
This one is also helpful as a comment in the PR but not in the resulting code

File: download_comments.py (hunk 1 above)
Comment (ericdaltman) [reply download_comments.py]: Works for me!

File: download_comments.py (hunk 2 above)
Comment (ericdaltman) [reply download_comments.py]: That should not have made it into the version I pushed. I made changes after originally writing that part and was sure I did a git add after. I'll check that out in my next commit.

File: download_posts.py (hunk 3 above)
Comment (ericdaltman) [reply download_posts.py]: That should definitely be rewritten. I wrote that bit of snark when I thought this would just be for me. I'll take a look at it.

File: download_posts.py (hunk 4 above)
Comment (ericdaltman) [reply download_posts.py]: Those were in the original code so I preserved them. If I recall those were inside the defintion of the dict being passed, so I figured they needed to be assigned? I'll take another look.

File: download_posts.py (hunk 5 above)
Comment (ericdaltman) [reply download_posts.py]: Likely not. Just missed those.

File: export.py (hunk 7 above)
Comment (ericdaltman) [reply export.py:33]: Not exactly sure what you're asking here?

File: export.py (hunk 8 above)
Comment (ericdaltman) [reply export.py]: Sure! I'll adjust.

File: download_posts.py (hunk 4 above)
Comment (ericdaltman) [reply download_posts.py:64]: Tested with removing these entries and they are required.

Comment (ericdaltman) [reply download_posts.py]: Got these confused with the ones inside the fetch_month_posts request. These can be removed.

File: export.py (hunk 7 above)
Comment (ericdaltman) [reply export.py]: Are you saying to capture the error from the requests, and feed to the except for reporting? Certainly could. Could even break it down into the various requests exceptions like timeouts and redirect issues.

File: export.py (hunk 9 above)
Comment (ericdaltman) [reply export.py]: Adjusted to state which cookie failed and an accompanying request to check connection and credentials.

File: export.py (hunk 10 above)
Comment (ericdaltman) [reply export.py:98]: Done and done!

Comment (ericdaltman) [reply export.py:98]: Went through and tried to recover other newlines and your original comments.

File: export.py (hunk 11 above)
Comment (ericdaltman) [reply export.py]: Done!

File: download_posts.py (hunk 3 above)
Comment (ericdaltman) [reply download_posts.py]: Adjusted to account for an error in either first or last year.

File: download_comments.py (hunk 2 above)
Comment (ericdaltman) [reply download_comments.py]: Cleaned up.

**Comment by ericdaltman** [comment]
Made requested adjustments aside from idea on a dedicated extraction of a cookie from a string which I like and can work on next.

File: export.py (hunk 6 above)
Comment (ericdaltman) [reply export.py]: Function added and pushed the new commit after testing. Please take a look and review.

File: export.py [hunk 12]
```diff
@@ -92,6 +158,19 @@ def json_to_markdown(json):
...
+    try:
+        header = response.headers.get('Set-Cookie')
+
+        if header:
+            return header.split(f'{cName}=')[1].split(';')[0]
+        else:
+            raise ValueError(f'Cookie {cName} not found in response.')
+
+    except Exception as e:
+        print(f"Error extracting cookie: {cName}. Error: {e}. Exiting...")
```
Comment (arty-name) [inline export.py]: That’s the kind of error handling I’d like to see in other `except` clauses: printing the actual exception

File: export.py [hunk 13]
```diff
@@ -92,6 +158,19 @@ def json_to_markdown(json):
...
+        sys.exit(1)
```
Comment (arty-name) [inline export.py]:
```suggestion
        sysexit(1)
```
I think this is a typo

File: export.py [hunk 14]
```diff
@@ -1,16 +1,82 @@
...
+if cookies.get('ljmastersession') is None:
+    print("Error: 'ljmastersession' cookie was not acquired.")
+    errCookie += 1
+if any(value is None for value in cookies.values()):
+    if errCookie == 1:
+        print("Only one of two required cookies acquired. Check connection and credentials. Exiting...")
+    else:
+        print("Failed to acquire both required cookies. Check connection and credentials. Exiting...")
+    sysexit(1)
+else:
```
Comment (arty-name) [inline export.py]: That’s why I wanted to see error handling in `get_cookie_value`: it makes this whole block obsolete, doesn’t it?

Comment (ericdaltman) [reply export.py]: Absolutely, as long as the cookie name is used in the error message then those messages can go.

File: export.py (hunk 13 above)
Comment (ericdaltman) [reply export.py]: Thank you!

File: export.py (hunk 12 above)
Comment (ericdaltman) [reply export.py]: Thanks!

Comment (ericdaltman) [reply export.py]: Updated the exception message on the line 31 as well:

```diff
@@ -29,3 +29,3 @@
- except:
...
-     print("Couldn't reach www.livejournal.com, please check internet connection. Exiting.")
+ except Exception as e:
...
+    print(f"Could not retrieve pre-connection cookie from www.livejournal.com. Error: {e}. Exiting.")
```

File: download_posts.py [hunk 15]
```diff
@@ -1,16 +1,27 @@
...
+import requests 
+from sys import exit as sysexit
 import xml.etree.ElementTree as ET
-from auth import cookies, headers
 
+# At time of script import, request range of years. We'll add 1 to the end year automatically.
+try:
+    startYear = int(input("Enter first year you want to export: "))
+except:
+    print("\nError with first year entered. Exiting...")
```
Comment (arty-name) [inline download_posts.py]:
```suggestion
except Exception as e:
    print(f"\nError with first year entered. Error: {e}. Exiting...")
```

File: download_posts.py [hunk 16]
```diff
@@ -1,16 +1,27 @@
...
+    sysexit(1)
 
-YEARS = range(2003, 2015)  # first to (last + 1)
+try:
+    endYear = int(input("Enter last year you want to export: "))
+except:
+    print("\nError with last year entered. Exiting...")
```
Comment (arty-name) [inline download_posts.py]:
```suggestion
except Exception as e:
    print(f"\nError with last year entered. Error: {e}. Exiting...")
```

**Review by arty-name — APPROVED** [review]
Very nice, Eric! I only have two remaining minor suggestions, and this PR is good to go!

File: download_posts.py (hunk 16 above)
Comment (ericdaltman) [reply download_posts.py]: Great! The other two files were next on my list. Incoming...

**Comment by ericdaltman** [comment]
So that should be everything. Had a small bug with function location which has been addressed. If you're happy with it, it is ready to merge.

**Comment by arty-name** [comment]
Yeah, I was wondering about the placement of this function. In Javascript, thanks to so-called hoisting, this would have  even worked. In any case the current code is cleaner. 

Thank you very much for your contribution, Eric! It made the tool much more user-friendly. I hope you liked the process!
